lib/lexer/lexer.cpp

```diff
@@ -0,0 +1,484 @@
+			return state_data.popState();
+		}
+		state_data.word.push_back(ch);
+		if (isDoubleQuote(ch)) {
+			return state_data.popState();
+		}
+		if (isBackslash(ch)) {
+			state = State::InnerBackslash;
+		} else if (isBackQuote(ch)) {
+			state = State::BackQuote;
+		} else if (isDollarSign(ch)) {
+			state = State::ExpansionStart;
+		}
+	}
+	// Nested state transition
+	state_data.pushState(State::DoubleQuote);
+	return state;
+}
+
+// preconditions:
+//	- backslash is already consumed and added to word
+//	- previous state is on the stack
+// Does NOT remove <newline> from the input stream
+Lexer::State Lexer::innerBackslashState() {
+	char ch = chars.consume();
+	if (ch == EOF) {
+		// todo: determine behavior
+		// throw std::runtime_error("encountered EOF in double quote backslash lexing");
+		return state_data.popState();
+	}
+	state_data.word.push_back(ch);
+	return state_data.popState();
+}
+
+Lexer::State Lexer::expansionStartState() {
+	switch (chars.peek()) {
+		case '{':
+			state_data.word.push_back(chars.consume());
+			return State::ParameterExpansion;
+		case '(':
+			state_data.word.push_back(chars.consume());
+			if (chars.peek() == '(') {
+				state_data.word.push_back(chars.consume());
+				return State::ArithmeticExpansion;
+			} else {
+				return State::CommandSubstitution;
+			}
+		default:
+			return state_data.popState();
+	}
+}
+
+// same as double quote, except it can also transition to single quote
+Lexer::State Lexer::nextNestedState(State current, char ch) {
+	if (ch == EOF) {
+		throw std::runtime_error("unexpected EOF in nested state");
+	} else if (isBackslash(ch)) {
+		return State::InnerBackslash;
+	} else if (isDoubleQuote(ch)) {
+		return State::DoubleQuote;
+	} else if (isSingleQuote(ch)) {
+		return State::SingleQuote;
+	} else if (isDollarSign(ch)) {
+		return State::ExpansionStart;
+	} else if (isBackQuote(ch)) {
+		return State::BackQuote;
+	}
+	return current;
+}
+
+Lexer::State Lexer::generalExpansionHandler(State state, char terminator) {
+	auto current = state;
+	while (current == state) {
+		char ch = chars.consume();
+		if (ch == EOF) {
+			return state_data.popState();
+		}
+		state_data.word.push_back(ch);
+		if (ch == terminator) {
+			return state_data.popState();
+		}
+		current = nextNestedState(current, ch);
+	}
+	state_data.pushState(state);
+	return current;
+}
+
+
+Lexer::State Lexer::parameterExpansionState() {
+	return generalExpansionHandler(state, '}');
+}
+
+Lexer::State Lexer::commandSubstitutionState() {
+	return generalExpansionHandler(state, ')');
+}
+
+// Requires some extra logic due to having multiple terminating characters
+Lexer::State Lexer::arithmeticExpansionState() {
+	auto state = State::ArithmeticExpansion;
+	while (state == State::ArithmeticExpansion) {
+		char ch = chars.consume();
+		if (ch == EOF) {
+			return state_data.popState();
+		}
+		state_data.word.push_back(ch);
+		if (ch == ')') {
+			if (chars.peek() == ')') {
```

Comment:
The logic for finding this end and disambiguating between a command substitution and a arithmetic expansion will also be required in the expander so it might be nice to make this into a util when that time comes.